### PR TITLE
Streaming events + UI rewrite

### PR DESCRIPTION
**Pull request by elijahbenizzy**
Big PR. See commits.

## Changes

 Changes:

1. Captured streaming events
2. Rewrote UI rendering of step list
3. Multiple other UI fixes
4. Fixes edge cases with recursive (burr-within-burr) applications

## How I tested this

Manually + tests + deployed to our staging area.

## Checklist

- [x] PR has an informative and human-readable title (this will be pulled into the release notes)
- [ ] Changes are limited to a single goal (no scope creep)
- [x] Code passed the pre-commit check & code is left cleaner/nicer than when first encountered.
- [x] Any _change_ in functionality is tested
- [x] New functions are documented (with a description, list of inputs, and expected output)
- [x] Placeholder code is flagged / future TODOs are captured in comments
- [ ] Project documentation has been updated if adding/changing functionality.

image
image
image

### CODE REVIEW

**Comment by elijahbenizzy** [comment]
TODO:
- [ ] Add unit tests
- [ ] Add documentation on apropriate functions/whatnot
- [ ] Add async version
- [ ] Add tracker functionality
    - [ ] Capture all events (first stream, etc...) 
    - [ ] Create data model
    - [ ] Log data
- [ ] Display in UI
    - [ ] Decide how
    - [ ] Render

**Comment by elijahbenizzy** [comment]
TODO:
- [ ] Fix tests
- [ ] Add any more tests that might be necessary
- [ ] Re-capture screenshots
- [ ] Regenerate demo data with all the new features

**Comment by skrawcz** [comment]
@elijahbenizzy this also closes #328  right? any other UI issues I had created?

File: examples/opentelemetry/application.py [hunk 1]
```diff
@@ -22,43 +21,49 @@
 @action(reads=[], writes=["chat_history", "prompt"])
 def process_prompt(state: State, prompt: str, __tracer: TracerFactory) -> Tuple[dict, State]:
     result = {"chat_item": {"role": "user", "content": prompt, "type": "text"}}
-    __tracer.log_attributes(prompt=prompt)
+    # __tracer.log_attributes(prompt=prompt)
```
Comment (skrawcz) [inline examples/opentelemetry/application.py:24]: commented out?

File: examples/opentelemetry/application.py [hunk 2]
```diff
@@ -82,43 +87,28 @@ def prompt_for_more(state: State) -> Tuple[dict, State]:
...
+    response = _query_openai(prompt=prompt, chat_history=chat_history_to_send, model=model)
+    return state.update(response={"content": response, "type": "text", "role": "assistant"})
 
 
 @action(reads=["prompt", "chat_history", "mode"], writes=["response"])
 def image_response(
     state: State, __tracer: TracerFactory, model: str = "dall-e-2"
 ) -> Tuple[dict, State]:
-    __tracer.log_attributes(model=model)
+    # __tracer.log_attributes(model=model)
```
Comment (skrawcz) [inline examples/opentelemetry/application.py:103]: commented out?

File: examples/opentelemetry/application.py [hunk 3]
```diff
@@ -210,11 +202,21 @@ def application_traceloop_as_otel_provider(
 
 if __name__ == "__main__":
     # Instrument OpenAI API
-    OpenAIInstrumentor().instrument()
+    # OpenAIInstrumentor().instrument()
```
Comment (skrawcz) [inline examples/opentelemetry/application.py:205]: commented out?

**Review by skrawcz — APPROVED** [review]
